SDL/Event.cu

```diff
@@ -599,20 +600,47 @@ __device__ int binary_search(
     return -1;
 }
 
-__global__ void moduleRangesKernel(uint16_t nLower, struct SDL::modules *modulesInGPU, struct SDL::hits *hitsInGPU)
+class moduleRangesKernel
 {
-    for ( int lowerIndex = blockIdx.x * blockDim.x + threadIdx.x; lowerIndex < nLower; lowerIndex += blockDim.x*gridDim.x )
+public:
+    ALPAKA_NO_HOST_ACC_WARNING
+    template<
+        typename TAcc,
+        typename TIdx>
+    ALPAKA_FN_ACC auto operator()(
+        TAcc const & acc,
+        struct SDL::modules *modulesInGPU,
+        struct SDL::hits *hitsInGPU,
+        TIdx const & nLowerModules) const
+    -> void
     {
-        uint16_t upperIndex = modulesInGPU->partnerModuleIndices[lowerIndex];
-        if (hitsInGPU->hitRanges[lowerIndex * 2] != -1 && hitsInGPU->hitRanges[upperIndex * 2] != -1)
+
+        TIdx const gridThreadIdx(alpaka::idx::getIdx<alpaka::Grid, alpaka::Threads>(acc)[0u]);
+        TIdx const threadElemExtent(alpaka::workdiv::getWorkDiv<alpaka::Thread, alpaka::Elems>(acc)[0u]);
```

Comment:
<img width="402" alt="image" src="https://user-images.githubusercontent.com/15358704/196046580-53da4533-8605-4b8d-a3dc-f3041d636ba0.png">
Mapping of Alpaka Abstraction to Hardware from [Alpaka – An Abstraction Library for Parallel Kernel Acceleration](https://arxiv.org/abs/1602.08477).

It seems that the elements do not have a clear mapping on a GPU. From the same paper:

> The element level represents an n-dimensional set of elements and unifies the data parallel capabilities of modern hardware architectures e.g. vectorization on thread level. [...] Alpaka therefore currently relies on compiler recognition of vectorizable code parts. Code is refactored in such a way that it includes primitive inner loops over a fixed number of elements.

Doing tests with a varying number of elements per thread, it seems that elements play a small role even on a GPU. However, I would possibly leave this level of abstraction as a next-to-leading order opportunity for optimization and work with the clear mappings up to threads for starters.

Reply:
there is a limited degree of vectorization possible with reduced precision on more recent architectures.
The paper is kind of old.
Maybe some effect can be seen in cases with half-precision data.